File: programs/inter-knot/src/state/bid.rs
use anchor_lang::prelude::*;

#[derive(AnchorSerialize, AnchorDeserialize, Clone, Copy, PartialEq, Eq, InitSpace)]
#[repr(u8)]
pub enum BidStatus {
    Active = 0,
    Selected = 1,
    Withdrawn = 2,
}

#[account]
#[derive(InitSpace)]
pub struct Bid {
    /// Which commission this bid is for
    pub commission_id: u64,
    /// Bidder's wallet
    pub executor: Pubkey,
    /// Bid price in USDC (6 decimals)
    pub price: u64,
    /// HTTP endpoint for x402 delivery
    #[max_len(128)]
    pub service_endpoint: String,
    /// Bid status
    pub status: BidStatus,
    /// Creation timestamp
    pub created_at: i64,
    /// PDA bump seed
    pub bump: u8,
}

File: sdk/src/types/bid.ts
import { PublicKey } from "@solana/web3.js";
import BN from "bn.js";

export type BidStatus = "active" | "selected" | "withdrawn";

export interface Bid {
  commissionId: number;
  executor: PublicKey;
  price: BN;
  serviceEndpoint: string;
  status: BidStatus;
  createdAt: BN;
  bump: number;
  /** The on-chain PDA address */
  address: PublicKey;
}

/** Parse the Anchor enum format { active: {} } → "active" */
export function parseBidStatus(raw: any): BidStatus {
  if (raw.active !== undefined) return "active";
  if (raw.selected !== undefined) return "selected";
  if (raw.withdrawn !== undefined) return "withdrawn";
  throw new Error(`Unknown bid status: ${JSON.stringify(raw)}`);
}
